src/Microsoft.AspNet.Mvc.Core/Internal/CopyOnWriteList.cs

```diff
@@ -0,0 +1,140 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNet.Mvc.Internal
+{
+    public class CopyOnWriteList<T> : IList<T>
```

Comment:
I'm also assuming that if this were a struct you'd lose all the benefits because all of the places that call through are expecting IList right?